Introduce forEachLayerAtPixel

tsauerwein opened the pull request:
This PR adds a method `forEachLayerAtPixel` to the map. This method calls a given callback for each layer that has a color value at a pixel. It is similar to `hasFeatureAtPixel` but this one can also be used on non-vector layers.

The main motivation is to use it together with `GetFeatureInfo` for WMS layers, like shown in this [example](http://tsauerwein.github.io/ol3/forEachLayerAtPixel/examples/getfeatureinfo-tile.html) (or with [renderer=webgl](http://tsauerwein.github.io/ol3/forEachLayerAtPixel/examples/getfeatureinfo-tile.html?renderer=webgl)).

    $(map.getViewport()).on('mousemove', function(evt) {
      var pixel = map.getEventPixel(evt.originalEvent);
      var hit = map.forEachLayerAtPixel(pixel, function(layer) {
        return true;
      });
      map.getTargetElement().style.cursor = hit ? 'pointer' : '';
    });

Closes #3097, cc @gberaudo

fredj commented:
works like a charm, thanks

gberaudo commented:
Is all new code covered by the two getfeatureinfo examples?
For example, it would be great to also test overlays.

I also added a comment about renaming functions taking a coordinate and asked if it would be possible to pass the pixel instead of the coordinate to avoid double conversions.

Otherwise, this PR looks pretty good, thank you @tsauerwein.

tsauerwein commented:
> Is all new code covered by the two getfeatureinfo examples?
For example, it would be great to also test overlays.

Locally, I made some changes to other example to be able to test. See https://github.com/tsauerwein/ol3/tree/forEachLayerAtPixel-testing It would be better to cover this by tests (see also #2923).

tsauerwein commented:
> Approximation errors might occur by converting pixel to coordinate then coordinate to pixel.

Fixed with 950ba7215567ab4d99ccf8a069964ca6120f546d

> The method is called forEachLayerAtPixel but takes a coordinate. In this case, it should be renamed to forEachLayerAtCoordinate.

Agree. I renamed all methods and also changed the method signatures, so that pixel/coordinate is the first argument.

Thanks for your remarks!

gberaudo commented:
Thank you for the good work @tsauerwein, +1 for merging.

elemoine commented:
Looking at this now…

elemoine commented on src/ol/map.js:
```diff
@@ -583,13 +583,50 @@ ol.Map.prototype.forEachFeatureAtPixel =
 
 
 /**
+ * Detect layers that have a colour value at a pixel on the viewport, and
```
I'm not sure I want to engage in this debate but s/colour/color?

elemoine commented:
Looks very good to me! This is a very nice addition.

elemoine commented:
I'd like to have this in v3.2.0, so please merge when you get a chance.

tsauerwein commented:
Merged, thanks for the reviews!
